src/main/scala/com/yahoo/tensorflowonspark/TFModel.scala

```diff
@@ -0,0 +1,348 @@
+        } else {
+          // array values
+          val numElements = shape.drop(1).map(_.toInt).product
+          dtype match {
+            case DataType.BOOL =>
+              val byteBuf = buf.asInstanceOf[ByteBuffer]
+              val byteArr = Array.ofDim[Byte](numElements)
+              byteBuf.get(byteArr, 0, numElements)
+              byteArr.map ( b => b == 1.toByte)
+            case DataType.DOUBLE =>
+              val doubleBuf = buf.asInstanceOf[DoubleBuffer]
+              val doubleArr = Array.ofDim[Double](numElements)
+              doubleBuf.get(doubleArr, 0, numElements)
+              doubleArr
+            case DataType.FLOAT =>
+              val floatBuf = buf.asInstanceOf[FloatBuffer]
+              val floatArr = Array.ofDim[Float](numElements)
+              floatBuf.get(floatArr, 0, numElements)
+              floatArr
+            case DataType.INT32 =>
+              val intBuf = buf.asInstanceOf[IntBuffer]
+              val intArr = Array.ofDim[Int](numElements)
+              intBuf.get(intArr, 0, numElements)
+              intArr
+            case DataType.INT64 =>
+              val longBuf = buf.asInstanceOf[LongBuffer]
+              val longArr = Array.ofDim[Long](numElements)
+              longBuf.get(longArr, 0, numElements)
+              longArr
+            case DataType.STRING =>
+              val byteBuf = buf.asInstanceOf[ByteBuffer]
+              val byteArr = Array.ofDim[Byte](buf.remaining())
+              byteBuf.get(byteArr)
+            case DataType.UINT8 =>
+              val byteBuf = buf.asInstanceOf[ByteBuffer]
+              val byteArr = Array.ofDim[Byte](buf.remaining())
+              byteBuf.get(byteArr)
+            case unsupportedType =>
+              Array()
+          }
+        }
+      }
+      result += Row(slices:_*)
+    }
+    result.toList
+  }
+
+  /**
+    * Transforms a Dataset of input data into an output DataFrame of predictions.
+    * Note: input columns are dropped during transform.
+    */
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    val spark = dataset.sparkSession
+
+    val inputColumns = this.getInputMapping.keys.toSeq
+    val inputTensorNames = this.getInputMapping.values
+    val outputTensorNames = this.getOutputMapping.keys.toSeq
+
+    val inputDF = dataset.select(inputColumns.head, inputColumns.tail: _*)
+    val inputSchema = inputDF.schema
+    val outputSchema = transformSchema(inputSchema)
+
+    val outputRDD = inputDF.rdd.mapPartitions { iter: Iterator[Row] =>
+      if (TFModel.model == null || TFModel.modelDir != this.getModel) {
+        // load model into a per-executor singleton reference, if needed.
+        TFModel.modelDir = this.getModel
+        TFModel.model = SavedModelBundle.load(this.getModel, this.getTag)
+        TFModel.graph = TFModel.model.graph
+        TFModel.sess = TFModel.model.session
+      }
+
+      var results = List.empty[Row]
+      val groupedIter = iter.grouped(this.getBatchSize)
```

Comment:
Do we require the size of each partition to be the multiples of batch size?

Reply:
No, shouldn't be required.  First dimension of most input tensors is typically `batch_size`, and `grouped` will produce a "remainder" batch if not a multiple.